README.md
# ss-clj

Clojure learn by doing --- Exchange Socket data from two server encrypted. Just like SS.

## Installation

![snapshot-runnable-jar](ss-clj-0.1.0-SNAPSHOT-standalone.jar)

## Usage

Server side:

    $ java -jar downloaded.jar --run-mode=rserver --port=[listen-port]  --password=[16-char-pwd]  --crypto-method=aes-cbc-128

Client side:

    $ java -jar downloaded.jar --run-mode=lserver --port=[listen-port] --rserver-addr=[server-ip] --rserver-port=[server-listening-port] --password=[16-char-pwd]  --crypto-method=aes-cbc-128

More:
   
    $ java -jar downloaded.jar -h


### Bugs

1. Some many children threads create now, no timeout watch yet.
2. client-server handshake not encrypto well




CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2015-12-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2015-12-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/ss-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/ss-clj/compare/0.1.0...0.1.1


test/ss_clj/core_test.clj
(ns ss-clj.core-test
  (:require [clojure.test :refer :all]
            [ss-clj.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/ss_clj/utils.clj
(ns ss-clj.utils
  (:gen-class)
  )

(defn byte2int [byte]
  (if (< byte 0)
    (int (+ 256 byte))
    (int byte)))

(defn int2byte [int]
  (unchecked-byte int))

(defn read-nbytes
  "read nbytes from input-stream, return byte-array"
  [input-stream count]
  (let [BUFF_SIZE 512
        byte-buffer (byte-array BUFF_SIZE)
        read-count (.read input-stream byte-buffer 0 count)]
    (byte-array read-count byte-buffer)))


(defn read-byte-to-int
  "read nbytes from input-stream, return as int"
  [input-stream]
  (byte2int (aget (read-nbytes input-stream 1) 0)))

(defn int2bytearray
  "bytearrays length is 2"
  [int]
  (byte-array [(unchecked-byte (bit-shift-right int 8)) (unchecked-byte int)]))

(defn bytearray->string
  ""
  [byte-array]
  (reduce str "" (map char byte-array)))

(defn bytearray->int
  "format bytearray to int"
  ([barray]
   (bytearray->int barray (count barray)))
  ([barray count]
   (if (= count 0)
     0
     (let [current_byte_int (byte2int (aget barray 0))
           current_shift (dec count)
           current_byte_result (bit-shift-left current_byte_int (* 8 current_shift))]
       (+ current_byte_result (bytearray->int (byte-array (rest barray)) (dec count)))))))

(defn bind-inputstream-to-outputstream
  "read is to buff, write buff to os.
  This Method is block when read or write "
  [is os whoami]
  (let [buff (byte-array 512)]
                                        ;(println whoami "start")
    (while (let [bytes-readed (.read is buff 0 512)
                 continue? (> bytes-readed 0)]
             (if continue?
               (do
                                        ; (println whoami bytes-readed)
                 (.write os buff 0 bytes-readed)
                 (.flush os)))
             continue?))
                                        ;(println whoami "end")
    ))


(defn build-connect-relay
  "build connect relay response package"
  [request redirect-result]
  (byte-array (+ 1 1 1 1 4 2)
              (concat [
                       (unchecked-byte 0x05)                   ; Ver
                       (unchecked-byte (:rep redirect-result))   ; rep
                       (unchecked-byte 0x00)                     ; rsv
                       (unchecked-byte 0x01) ; atyp -> ipv4, always ipv4
                       ]
                      [0 0 0 0]         ;addr
                      (int2bytearray (let [socket (:socket redirect-result)]
                                       (if (= nil socket)
                                         0
                                         (.getLocalPort socket)))))))


src/ss_clj/crypto.clj
(ns ss-clj.crypto
  (:gen-class)
  (:use ss-clj.utils)
  (import (java.io ByteArrayInputStream FileInputStream FileOutputStream File ByteArrayOutputStream)
          (javax.crypto Cipher)
          (javax.crypto.spec SecretKeySpec IvParameterSpec)))

(def IV (byte-array 16 '(8)))

(defn encrypto
  ""
  [data key]
  (let [cipher (Cipher/getInstance "AES/CBC/NoPadding" "SunJCE")
        keySpec (SecretKeySpec. (.getBytes key "utf-8") "AES")
        ivp (IvParameterSpec. IV)
        ]
    (do
      (.init cipher Cipher/ENCRYPT_MODE keySpec ivp)
      (.doFinal cipher data))))


(defn decrypto
  ""
  [data key]
  (let [cipher (Cipher/getInstance "AES/CBC/NOPadding" "SunJCE")
        keySpec (SecretKeySpec. (.getBytes key "utf-8") "AES")
        ivp (IvParameterSpec. IV)
        ]
    (do
      (.init cipher Cipher/DECRYPT_MODE keySpec ivp)
      ;(println cipher data)
      (.doFinal cipher data))))

(defn wraper-encrypto-inputstream-2-outputstream
  [is os whoami key]
  (let [package-length 512
        resver-length 2
        payload-length (- package-length resver-length)
        buff (byte-array package-length)
        ]
    (while (let [readed-count (.read is buff resver-length payload-length)
                 readed-count-raw-array (int2bytearray readed-count)
                 continue? (> readed-count 0)
                 ]
             (if continue?
               (do
                 ;; copy resver length
                 (aset-byte buff 0 (aget readed-count-raw-array 0))
                 (aset-byte buff 1 (aget readed-count-raw-array 1))
                 (.write os (encrypto buff key))
                 ))
             continue?))))

(defn blockread-inputstream-align
  [is buff start end]
  (let [readed-count (.read is buff start (- end start))
        continue? (and (> readed-count 0) (< (+ start readed-count) end))]
    (if continue?
      (+ readed-count (blockread-inputstream-align is buff (+ start readed-count) end))
      readed-count)))

(defn unwraper-encrypto-inputstream-2-outputstream
  [is os whoami key]
  (let [package-length 512
        resver-length 2
        payload-length (- package-length resver-length)
        buff (byte-array package-length)
        ]
    (while (let [readed-count (blockread-inputstream-align is buff 0 package-length)
                 after-de-buff (decrypto buff  key)
                 payload-array (byte-array (take resver-length after-de-buff))
                 payload-size (bytearray->int payload-array)
                 continue? (> readed-count 0)
                 ]
             (if continue?
               (do
                 ;; copy resver length
                 (.write os  after-de-buff resver-length payload-size)))
             continue?))))

(def crypto-methods
  {"plain" {:init_fn #()
            :wrapper #(apply bind-inputstream-to-outputstream (take 3 %&))
            :unwraper #(apply bind-inputstream-to-outputstream (take 3 %&))}
   "aes-cbc-128" {:init_fn #(println "no implement")
                  :wrapper wraper-encrypto-inputstream-2-outputstream
                  :unwraper unwraper-encrypto-inputstream-2-outputstream}})

(defn test-wrap-unwrap
  [dataout]
  (let [data dataout
        TEST_KEY "0123456789123456"]
    (with-open [before-en-is (ByteArrayInputStream. (.getBytes data "utf-8"))
                after-en-os (FileOutputStream. (File. "/home/zhujj/tmp2"))]
      (wraper-encrypto-inputstream-2-outputstream before-en-is after-en-os "nothing" TEST_KEY))
    (with-open [before-de-is (FileInputStream. (File. "/home/zhujj/tmp2"))
                after-de-os (FileOutputStream. (File. "/home/zhujj/tmp3"))]
      (unwraper-encrypto-inputstream-2-outputstream before-de-is after-de-os "nothing" TEST_KEY))
    (slurp "/home/zhujj/tmp3")))


src/ss_clj/local_server.clj
(ns ss-clj.local_server
  (:gen-class)
  (:use ss-clj.utils)
  (:require [ss-clj.crypto :as crypto])
  (:import (java.net ServerSocket Socket InetSocketAddress UnknownHostException)
           (java.io BufferedReader InputStreamReader InputStream ByteArrayInputStream PrintWriter)))


(def println #(apply clojure.core/println "LOCAL: " %&))

(def local-config nil)

(defn process-sock5-handshake-request
  "process sock5 handshake-request into a map, like:
  {:version n, :nmethods n, :methods '(& n)}"
  [input-stream]
  (let [parsed-request (assoc {} :version (read-byte-to-int input-stream))]
    (if (= (:version parsed-request) 5)
      (let [nmethods (read-byte-to-int input-stream)
            methods (read-nbytes input-stream nmethods)]
        (assoc parsed-request :nmethods nmethods :methods (seq methods)))
      parsed-request)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn test-process-sock5-handshake-request
  "Test Function"
  []
  (assert ; test two method
   (= (let [zbuff (byte-array [(byte 0x05) (byte 0x02) (byte 0x00) (byte 0x01)])]
        (with-open [test-is (ByteArrayInputStream. zbuff)]
          (process-sock5-handshake-request test-is)))
      {:version 5 :nmethods 2 :methods '(0 1)}))
  (assert ; test under 5
   (= (let [zbuff (byte-array [(byte 0x03) (byte 0x02) (byte 0x00) (byte 0x01)])]
        (with-open [test-is (ByteArrayInputStream. zbuff)]
          (process-sock5-handshake-request test-is)))
      {:version 3})))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def accept-request "only support unauthorized"
  (byte-array [(byte 0x05) (byte 0x00)]))
(def deny-request "deny request"
  (byte-array [(byte 0x05) (unchecked-byte 0xff)]))

(defn parse-dst-addr
  "parse dst addr accord atyp"
  [atyp is]
  (cond
    (= atyp 0x01)                       ;parse ipv4
    (let [addr (read-nbytes is 4)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x03)                       ;parse domain name
    (let [domain-length (read-byte-to-int is)
          addr (read-nbytes is domain-length)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x04)                       ;parse ipv6
    (let [addr (read-nbytes is 6)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))))

(defn parse-sock5-request
  "process sock5 request into map, like"
  [is]
  (let [version (read-byte-to-int is)
        cmd (read-byte-to-int is)
        rsv (read-byte-to-int is)
        atyp (read-byte-to-int is)
        dst-addr (parse-dst-addr atyp is)]
    (assoc {}
           :version version
           :cmd cmd
           :rsv rsv
           :atyp atyp
           :dst dst-addr)))

(def method-not-supported
  (byte-array [
               (unchecked-byte 0x05)    ;Ver
               (unchecked-byte 0x07)    ;rep
               (unchecked-byte 0x00)    ;rsv
               (unchecked-byte 0x01)    ;atype
               (unchecked-byte 0x00)    ;addr0
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;addr3
               (unchecked-byte 0x00)    ;port0
               (unchecked-byte 0x00)    ;port1
               ]))

(defn open-connect-with-remote-server
  "open remote proxy socket, return map contain socket and reply code
  bytes format is:
Req:
| open-assecss-key        | method | atyp | addr | port |
|      16                 | 1      |  1   | n    | 2    |

Resp:
| 4length | result      |
| 4       | n (4length) |
"
  [open-access-key method atyp dst-addr dst-port]
  (try
    (let [remote-socket (Socket. (:rserver-addr local-config) (:rserver-port local-config))
          is (.getInputStream remote-socket)
          os (.getOutputStream remote-socket)
          domain-addr-prefix-barray (if (= atyp 3) (byte-array 1 [(unchecked-byte (count dst-addr))]) [])
          req-package (byte-array (concat (if (= "plain" (:crypto-method local-config)) (byte-array 16)
                                              (crypto/encrypto open-access-key (:password local-config)))
                                          (byte-array [(int2byte method)])
                                          (byte-array [(int2byte atyp)])
                                          domain-addr-prefix-barray
                                          dst-addr
                                          dst-port))
          read-buff (byte-array 4)]
      ;(println "query remote : "  (seq req-package))
      (.write os req-package)
      (.read is read-buff 0 4)
      (let [resp-length (bytearray->int read-buff)
            resp-read-buff (byte-array resp-length)
            resp-payload-length (.read is resp-read-buff 0 resp-length)
            resp-rsp-code (bytearray->int resp-read-buff)]
        ;(println "response of remote : "  resp-length (seq resp-read-buff))
        ; TODO parse buff. For now we just translate to int
        (if (and (= resp-payload-length resp-length) ; desire length == read length && code == 0x00
                 (= resp-rsp-code 0x0))
          {:socket remote-socket :rep 0x00}
          {:socket nil :rep 0x1})))
    (catch UnknownHostException e
      {:socket nil :rep 0x03})
    (catch Exception e
      (do (println (.getMessage e)) {:socket nil :rep 0x01}))))

(defn process-sock5-request
  ""
  [request socket is os]
  (let [cmd (:cmd request)]
    (cond (= cmd 0x1)                   ; Connect
          (let [dst-addr (get-in request [:dst :dst-addr])
                dst-port (get-in request [:dst :dst-port])
                method 1
                atyp (:atyp request)
                password (:password local-config)
                open-access-key (.getBytes password "utf-8")
                redirect-socket-result (open-connect-with-remote-server open-access-key method atyp dst-addr dst-port)
                ]
            ;(println "connect done")
            ;(println redirect-socket-result)
            (.write os (build-connect-relay request redirect-socket-result))
            (if (= 0 (:rep redirect-socket-result))
              ;; loop read & write is streams
              (with-open [proxy-socket (:socket redirect-socket-result)
                          client-socket socket
                          pis (.getInputStream proxy-socket)
                          pos (.getOutputStream proxy-socket)
                          cis is
                          cos os]
                ;(println "start proxy loop: " (.getLocalPort proxy-socket))
                (doto (Thread. #((get-in crypto/crypto-methods [(:crypto-method local-config) :unwraper])  pis cos "pis->cos" password)) (.start))
                ((get-in crypto/crypto-methods [(:crypto-method local-config) :wrapper]) cis pos "cis->pos" password)

                ;(println "end proxy loop")
                )
              ))
          (= cmd 0x2)                   ; Bind
          (.write os method-not-supported)
          (= cmd 0x3)                   ; UDP Accociate
          (.write os method-not-supported))))


(defn dump-sock5-request-map
  "print request map"
  [request-map]
  (let [dst-addr (get-in request-map [:dst :dst-addr] nil)
        dst-port (get-in request-map [:dst :dst-port] nil)
        parsed-dst-addr (bytearray->string dst-addr)
        parsed-dst-port (bytearray->int dst-port)]
    (println (assoc-in (assoc-in request-map [:dst :dst-addr] parsed-dst-addr)
               [:dst :dst-port] parsed-dst-port))))

(defn handle-client-request
  ""
  [socket]
  ;(println "new client request start\n")
  (let [input-stream (.getInputStream socket)
        output-stream (.getOutputStream socket)
        request (process-sock5-handshake-request input-stream)]
    ;(println request)
    (if (= (:version request) 5)
      (do (.write output-stream accept-request)
          (let [request (parse-sock5-request input-stream)]
            ;(dump-sock5-request-map request)
            (process-sock5-request request socket input-stream output-stream)))
      (.write output-stream deny-request))))

(defn init-config
  [config-map]
  (def local-config config-map))


src/ss_clj/remote_server.clj
(ns ss-clj.remote_server
  (:gen-class)
  (:use ss-clj.utils)
  (:require [ss-clj.crypto :as crypto])
  (:import (java.net ServerSocket Socket InetSocketAddress UnknownHostException)
           (java.io BufferedReader InputStreamReader InputStream ByteArrayInputStream PrintWriter)))


(def println #(apply clojure.core/println "SERVER: " %&))
(def local-config {})

(declare parse-dst-addr)

(defn process-openaccess-handshake-request
  "process openaccess handshake-request into a map, like:
  {:open-accesss-key n, :methods n, :atyp n, :dst {:dst-addr :dst-port}}"
  [input-stream]
  (let [open-access-key (read-nbytes input-stream 16) ;open-access-key
        method          (read-byte-to-int input-stream)  ;method
        atyp            (read-byte-to-int input-stream) ;atyp
        addr            (parse-dst-addr atyp input-stream) ; addr
        ]
    {:open-access-key open-access-key :cmd method :atyp atyp :dst addr}))


(def deny-request "deny request"
  (byte-array [0 0 0 1; 4-length
               1; deny
               ]))

(defn parse-dst-addr
  "parse dst addr accord atyp"
  [atyp is]
  (cond
    (= atyp 0x01)                       ;parse ipv4
    (let [addr (read-nbytes is 4)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x03)                       ;parse domain name
    (let [domain-length (read-byte-to-int is)
          addr (read-nbytes is domain-length)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x04)                       ;parse ipv6
    (let [addr (read-nbytes is 6)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))))

(def method-not-supported
  (byte-array [
               (unchecked-byte 0x05)    ;Ver
               (unchecked-byte 0x07)    ;rep
               (unchecked-byte 0x00)    ;rsv
               (unchecked-byte 0x01)    ;atype
               (unchecked-byte 0x00)    ;addr0
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;addr3
               (unchecked-byte 0x00)    ;port0
               (unchecked-byte 0x00)    ;port1
               ]))

(defn open-connect
  "return map contain socket and reply code"
  [host port]
  (try
    {:socket (Socket. host port) :rep 0x00}
    (catch UnknownHostException e
      {:socket nil :rep 0x03})
    (catch Exception e
      {:socket nil :rep 0x01})))

(defn build-open-access-reply
  ""
  [request redirect-socket-result]
  (byte-array [0 0 0 1 (unchecked-byte (:rep redirect-socket-result))]))

(defn process-validate-open-access-request
  ""
  [request socket is os]
  (let [cmd (:cmd request)]
    (cond (= cmd 0x1)                   ; Connect
          (do
            (let [dst-addr (get-in request [:dst :dst-addr])
                  dst-port (get-in request [:dst :dst-port])
                  host (bytearray->string dst-addr)
                  port (bytearray->int    dst-port)
                  redirect-socket-result (open-connect host port)
                  password (:password local-config)]
              ;(println host)
              (.write os (build-open-access-reply request redirect-socket-result))
              (if (= 0 (:rep redirect-socket-result))
                ;; loop read & write is streams
                (with-open [proxy-socket (:socket redirect-socket-result)
                            client-socket socket
                            pis (.getInputStream proxy-socket)
                            pos (.getOutputStream proxy-socket)
                            cis is
                            cos os]
                  ;(println "start proxy loop: " (.getLocalPort proxy-socket))
                  ;; (doto (Thread. #(bind-inputstream-to-outputstream pis cos "pis->cos")) (.start))
                  ;; (bind-inputstream-to-outputstream cis pos "cis->pos")
                  (doto (Thread. #((get-in crypto/crypto-methods [(:crypto-method local-config) :wrapper])  pis cos "pis->cos" password)) (.start))
                  ((get-in crypto/crypto-methods [(:crypto-method local-config) :unwraper]) cis pos "cis->pos" password)
                  ;(println "end proxy loop")
                  )
                )
              ))
          (= cmd 0x2)                   ; Bind
          (.write os method-not-supported)
          (= cmd 0x3)                   ; UDP Accociate
          (.write os method-not-supported))))

(defn handle-client-request
  ""
  [socket]
  ;(println "new client request start\n")
  (let [input-stream (.getInputStream socket)
        output-stream (.getOutputStream socket)
        request (process-openaccess-handshake-request input-stream)]
    ;(println request)
    (if (if (= "plain" (:crypto-method local-config)) (= (take 16 (repeat 0)) (seq (:open-access-key request)))
            (=                          ; test open-access-key
             (String. (crypto/decrypto (:open-access-key request) (:password local-config)))
             (:password local-config)))             
      (process-validate-open-access-request request socket input-stream output-stream)
      (.write output-stream deny-request))
    ))

(defn init-config
  [config-map]
  (def local-config config-map))


src/ss_clj/socks5_server.clj
(ns ss-clj.socks5_server
  (:gen-class)
  (:use (ss-clj utils))
  (:import (java.net ServerSocket Socket InetSocketAddress UnknownHostException)
           (java.io BufferedReader InputStreamReader InputStream ByteArrayInputStream PrintWriter)))

(defn process-sock5-handshake-request
  "process sock5 handshake-request into a map, like:
  {:version n, :nmethods n, :methods '(& n)}"
  [input-stream]
  (let [parsed-request (assoc {} :version (read-byte-to-int input-stream))]
    (if (= (:version parsed-request) 5)
      (let [nmethods (read-byte-to-int input-stream)
            methods (read-nbytes input-stream nmethods)]
        (assoc parsed-request :nmethods nmethods :methods (seq methods)))
      parsed-request)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn test-process-sock5-handshake-request
  "Test Function"
  []
  (assert ; test two method
   (= (let [zbuff (byte-array [(byte 0x05) (byte 0x02) (byte 0x00) (byte 0x01)])]
        (with-open [test-is (ByteArrayInputStream. zbuff)]
          (process-sock5-handshake-request test-is)))
      {:version 5 :nmethods 2 :methods '(0 1)}))
  (assert ; test under 5
   (= (let [zbuff (byte-array [(byte 0x03) (byte 0x02) (byte 0x00) (byte 0x01)])]
        (with-open [test-is (ByteArrayInputStream. zbuff)]
          (process-sock5-handshake-request test-is)))
      {:version 3})))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(def accept-request "only support unauthorized"
  (byte-array [(byte 0x05) (byte 0x00)]))
(def deny-request "deny request"
  (byte-array [(byte 0x05) (unchecked-byte 0xff)]))

(defn parse-dst-addr
  "parse dst addr accord atyp"
  [atyp is]
  (cond
    (= atyp 0x01)                       ;parse ipv4
    (let [addr (read-nbytes is 4)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x03)                       ;parse domain name
    (let [domain-length (read-byte-to-int is)
          addr (read-nbytes is domain-length)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))
    (= atyp 0x04)                       ;parse ipv6
    (let [addr (read-nbytes is 6)
          port (read-nbytes is 2)]
      (assoc {} :dst-addr addr :dst-port port))))

(defn parse-sock5-request
  "process sock5 request into map, like"
  [is]
  (let [version (read-byte-to-int is)
        cmd (read-byte-to-int is)
        rsv (read-byte-to-int is)
        atyp (read-byte-to-int is)
        dst-addr (parse-dst-addr atyp is)]
    (assoc {}
           :version version
           :cmd cmd
           :rsv rsv
           :atyp atyp
           :dst dst-addr)))

(def method-not-supported
  (byte-array [
               (unchecked-byte 0x05)    ;Ver
               (unchecked-byte 0x07)    ;rep
               (unchecked-byte 0x00)    ;rsv
               (unchecked-byte 0x01)    ;atype
               (unchecked-byte 0x00)    ;addr0
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;
               (unchecked-byte 0x00)    ;addr3
               (unchecked-byte 0x00)    ;port0
               (unchecked-byte 0x00)    ;port1
               ]))

(defn open-connect
  "return map contain socket and reply code"
  [host port]
  (try
    {:socket (Socket. host port) :rep 0x00}
    (catch UnknownHostException e
      {:socket nil :rep 0x03})
    (catch Exception e
      {:socket nil :rep 0x01})))

(defn process-sock5-request
  ""
  [request socket is os]
  (let [cmd (:cmd request)]
    (cond (= cmd 0x1)                   ; Connect
          (do
            (let [dst-addr (get-in request [:dst :dst-addr])
                  dst-port (get-in request [:dst :dst-port])
                  host (bytearray->string dst-addr)
                  port (bytearray->int    dst-port)
                  redirect-socket-result (open-connect host port)]
              (println host)
              (.write os (build-connect-relay request redirect-socket-result))
              (if (= 0 (:rep redirect-socket-result))
                ;; loop read & write is streams
                (with-open [proxy-socket (:socket redirect-socket-result)
                            client-socket socket
                            pis (.getInputStream proxy-socket)
                            pos (.getOutputStream proxy-socket)
                            cis is
                            cos os]
                  (println "start proxy loop: " (.getLocalPort proxy-socket))
                  (doto (Thread. #(bind-inputstream-to-outputstream pis cos "pis->cos")) (.start))
                  (bind-inputstream-to-outputstream cis pos "cis->pos")
                  (println "end proxy loop"))
                )
              ))
          (= cmd 0x2)                   ; Bind
          (.write os method-not-supported)
          (= cmd 0x3)                   ; UDP Accociate
          (.write os method-not-supported))))


(defn dump-sock5-request-map
  "print request map"
  [request-map]
  (let [dst-addr (get-in request-map [:dst :dst-addr] nil)
        dst-port (get-in request-map [:dst :dst-port] nil)
        parsed-dst-addr (bytearray->string dst-addr)
        parsed-dst-port (bytearray->int dst-port)]
    (println (assoc-in (assoc-in request-map [:dst :dst-addr] parsed-dst-addr)
               [:dst :dst-port] parsed-dst-port))))

(defn handle-client-request
  ""
  [socket]
  (println "new client request start\n")
  (let [input-stream (.getInputStream socket)
        output-stream (.getOutputStream socket)
        request (process-sock5-handshake-request input-stream)]
    (println request)
    (if (= (:version request) 5)
      (do (.write output-stream accept-request)
          (let [request (parse-sock5-request input-stream)]
            (dump-sock5-request-map request)
            (process-sock5-request request socket input-stream output-stream)))
      (.write output-stream deny-request))))


src/ss_clj/core.clj
(ns ss-clj.core
  (:gen-class)
  (:use (ss-clj utils))
  (:require [ss-clj.socks5_server :as s5server]
            [ss-clj.local_server :as lserver]
            [ss-clj.remote_server :as rserver]
            [clojure.tools.cli :as cli])
  (:import (java.net ServerSocket Socket InetSocketAddress UnknownHostException)
           (java.io BufferedReader InputStreamReader InputStream ByteArrayInputStream PrintWriter)))

(defn println-socket
  [socket]
  (let [buff (byte-array 8)
        is (.getInputStream socket)]
    (while (let [readcount (.read is buff 0 8)
                 continue? (> readcount 0)]
             (if continue?
               (println "println-socket---" (seq buff)))
             continue?))))

(def socks5-proxy-server s5server/handle-client-request)
(def local-server #(do (lserver/init-config %) lserver/handle-client-request))
(def remote-server #(do (rserver/init-config %) rserver/handle-client-request))

(defn socket-server
  [handler port]
  (with-open [server-socket (ServerSocket. port)]
    (println "Listening Port: " (.getLocalPort server-socket))
    (while true
      (let [socket (.accept server-socket)]
        (doto (Thread. #(handler socket)) (.start))))))

(defn dev-test
  ""
  [commandline-config]
  (let [ 
        remote-server-config {:crypto-method (:crypto-method commandline-config)
                              :password (:password commandline-config)}
        local-server-config {:rserver-addr "127.0.0.1"
                             :rserver-port 8009
                             :crypto-method (:crypto-method commandline-config)
                             :password (:password commandline-config)}]
    (doto (Thread. #(socket-server (remote-server remote-server-config) 8009)) (.start))
    (socket-server (local-server local-server-config) 8008)))

(def cli-options
  ;; An option with a required argument
  [["-m" "--run-mode MODE" "MODEs: socks5-proxy rserver lserver dev-test"
    :default "socks5-proxy"
    :parse-fn #(str %)
    :validate [#(or (= % "socks5-proxy") (= % "lserver") (= % "rserver") (= % "dev-test")) "Must one of: socks5-proxy rserver lserver"]]
   [nil "--rserver-addr RServer-ADDR" "only work when mode==lserver"
    :default "127.0.0.1"
    ]
   [nil "--rserver-port RServer-Port" "only work when mode==lserver"
    :default 8009
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]
    ]
   [nil "--crypto-method Method" "MODEs: plain aes-cbc-128"
    :default "plain"
    :parse-fn #(str %)
    :validate [#(or (= % "plain") (= % "aes-cbc-128")) "Must one of: plain aes-cbc-128"]]
   [nil "--password PASSWORD" "password, now it must 16-bytes"
    :default "0123456789123456"
    :parse-fn #(str %)
    :validate [#(= (count %) 16) "Must 16-bytes"]]
   ["-p" "--port PORT" "Port number"
    :default 8008
    :parse-fn #(Integer/parseInt %)
    :validate [#(< 0 % 0x10000) "Must be a number between 0 and 65536"]]
   ;; A non-idempotent option
   ["-v" nil "Verbosity level"
    :id :verbosity
    :default 0
    :assoc-fn (fn [m k _] (update-in m [k] inc))]
   ;; A boolean option defaulting to nil
   ["-h" "--help"]])



(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (let [argument (cli/parse-opts args cli-options)
        mode (get-in argument [:options :run-mode])
        port (get-in argument [:options :port])]
    (cond
      (:errors argument) ; error
      (println (:errors argument))
          
      (get-in argument [:options :help])
      (println (:summary argument)) ; help

      (= mode "dev-test")
      (do (println mode)
          (dev-test (:options argument)))

      (= mode "socks5-proxy")                            ; socks5-proxy
      (do (println mode) (socket-server socks5-proxy-server port))

      (= mode "rserver")                ;rserver
      (do (println mode)
          (socket-server (remote-server (:options argument)) port))

      (= mode "lserver")                ;lserver
      (do (println mode)
          (socket-server (local-server (:options argument)) port))
          ))
  )
